File: android/src/main/java/it/hubagency/spatialscan/UnlinkedOpeningStore.kt
package it.hubagency.spatialscan

import android.content.Context
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.UUID

/**
 * Registro persistente delle aperture interne non ancora collegate bilateralmente.
 * Salvato in filesDir/hub_unlinked_openings.json.
 *
 * Ciclo di vita di una entry:
 *  1. Viene creata quando l'utente conferma un'apertura interna con status PENDING
 *     (sourceRoomId viene popolato dopo il salvataggio del RoomRecord).
 *  2. Viene rimossa quando l'utente collega quella apertura durante la scan
 *     di un ambiente adiacente (status diventa LINKED bilateralmente).
 */
data class UnlinkedOpening(
    val id:             String,   // UUID entry nel registro
    val sourceRoomId:   String,   // UUID del RoomRecord sorgente
    val sourceRoomName: String,   // nome leggibile ("Corridoio")
    val openingId:      String,   // OpeningModel.id nell'ambiente sorgente
    val kind:           OpeningKind,
    val width:          Float,
    val height:         Float,
    val bottom:         Float,
    val wallIndex:      Int,      // indice parete nell'ambiente sorgente (per etichetta)
    val customLabel:    String = ""  // etichetta utente es. "cucina", "bagno"
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("id",             id)
        put("sourceRoomId",   sourceRoomId)
        put("sourceRoomName", sourceRoomName)
        put("openingId",      openingId)
        put("kind",           kind.name)
        put("width",          width)
        put("height",         height)
        put("bottom",         bottom)
        put("wallIndex",      wallIndex)
        put("customLabel",    customLabel)
    }

    companion object {
        fun fromJson(obj: JSONObject): UnlinkedOpening? = runCatching {
            UnlinkedOpening(
                id             = obj.getString("id"),
                sourceRoomId   = obj.getString("sourceRoomId"),
                sourceRoomName = obj.getString("sourceRoomName"),
                openingId      = obj.getString("openingId"),
                kind           = OpeningKind.valueOf(obj.getString("kind")),
                width          = obj.optDouble("width",  0.80).toFloat(),
                height         = obj.optDouble("height", 2.10).toFloat(),
                bottom         = obj.optDouble("bottom", 0.00).toFloat(),
                wallIndex      = obj.optInt("wallIndex", 0),
                customLabel    = obj.optString("customLabel", "")
            )
        }.getOrNull()
    }
}

object UnlinkedOpeningStore {

    private const val TAG      = "UnlinkedOpeningStore"
    private const val FILENAME = "hub_unlinked_openings.json"

    fun add(context: Context, entry: UnlinkedOpening) {
        try {
            val list = loadAll(context).toMutableList()
            list.add(entry)
            writeAll(context, list)
            Log.d(TAG, "added opening id=${entry.id} from '${entry.sourceRoomName}'")
        } catch (e: Exception) {
            Log.e(TAG, "add failed: ${e.message}", e)
        }
    }

    fun remove(context: Context, id: String) {
        try {
            val list = loadAll(context).filter { it.id != id }
            writeAll(context, list)
            Log.d(TAG, "removed opening id=$id")
        } catch (e: Exception) {
            Log.e(TAG, "remove failed: ${e.message}", e)
        }
    }

    fun loadAll(context: Context): List<UnlinkedOpening> {
        return try {
            val f = File(context.filesDir, FILENAME)
            if (!f.exists()) return emptyList()
            val arr = JSONArray(f.readText())
            (0 until arr.length()).mapNotNull {
                runCatching { UnlinkedOpening.fromJson(arr.getJSONObject(it)) }.getOrNull()
            }
        } catch (e: Exception) {
            Log.e(TAG, "loadAll failed: ${e.message}", e)
            emptyList()
        }
    }

    private fun writeAll(context: Context, entries: List<UnlinkedOpening>) {
        val arr = JSONArray()
        entries.forEach { arr.put(it.toJson()) }
        File(context.filesDir, FILENAME).writeText(arr.toString())
    }
}

File: ios/Plugin/UnlinkedOpeningStore.swift
import Foundation

/// Registro persistente delle aperture interne non ancora collegate bilateralmente.
/// Porta iOS di UnlinkedOpeningStore.kt.
/// Persistenza: Documents/hub_unlinked_openings.json
struct UnlinkedOpening {
    let id:             String   // UUID entry nel registro
    var sourceRoomId:   String   // UUID del RoomRecord sorgente (popolato dopo il save)
    let sourceRoomName: String   // nome leggibile ("Corridoio")
    let openingId:      String   // OpeningModel.id nell'ambiente sorgente
    let kind:           OpeningKind
    let width:          Float
    let height:         Float
    let bottom:         Float
    let wallIndex:      Int
    let customLabel:    String

    func withSourceRoomId(_ newId: String) -> UnlinkedOpening {
        UnlinkedOpening(id: id, sourceRoomId: newId, sourceRoomName: sourceRoomName,
                        openingId: openingId, kind: kind, width: width, height: height,
                        bottom: bottom, wallIndex: wallIndex, customLabel: customLabel)
    }

    func toDict() -> [String: Any] {
        ["id": id, "sourceRoomId": sourceRoomId, "sourceRoomName": sourceRoomName,
         "openingId": openingId, "kind": kind.rawValue,
         "width": width, "height": height, "bottom": bottom,
         "wallIndex": wallIndex, "customLabel": customLabel]
    }

    static func fromDict(_ d: [String: Any]) -> UnlinkedOpening? {
        guard let id            = d["id"]             as? String,
              let sourceRoomId  = d["sourceRoomId"]   as? String,
              let sourceRoomName = d["sourceRoomName"] as? String,
              let openingId     = d["openingId"]       as? String,
              let kindStr       = d["kind"]            as? String,
              let kind          = OpeningKind(rawValue: kindStr) else { return nil }
        return UnlinkedOpening(
            id:             id,
            sourceRoomId:   sourceRoomId,
            sourceRoomName: sourceRoomName,
            openingId:      openingId,
            kind:           kind,
            width:          (d["width"]       as? NSNumber)?.floatValue ?? 0.80,
            height:         (d["height"]      as? NSNumber)?.floatValue ?? 2.10,
            bottom:         (d["bottom"]      as? NSNumber)?.floatValue ?? 0.00,
            wallIndex:      (d["wallIndex"]   as? Int) ?? 0,
            customLabel:    (d["customLabel"] as? String) ?? ""
        )
    }
}

final class UnlinkedOpeningStore {

    static let shared = UnlinkedOpeningStore()
    private init() {}

    private let filename = "hub_unlinked_openings.json"

    func add(_ entry: UnlinkedOpening) {
        var list = loadAll()
        list.append(entry)
        writeAll(list)
    }

    func remove(id: String) {
        let list = loadAll().filter { $0.id != id }
        writeAll(list)
    }

    func loadAll() -> [UnlinkedOpening] {
        guard let url  = fileURL,
              let data = try? Data(contentsOf: url),
              let arr  = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
        else { return [] }
        return arr.compactMap { UnlinkedOpening.fromDict($0) }
    }

    func clear() {
        writeAll([])
    }

    private func writeAll(_ list: [UnlinkedOpening]) {
        guard let url  = fileURL,
              let data = try? JSONSerialization.data(withJSONObject: list.map { $0.toDict() })
        else { return }
        try? data.write(to: url)
    }

    private var fileURL: URL? {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first?
            .appendingPathComponent(filename)
    }
}
